### PR TITLE
[Stratconn-2441] | Combined Braze custom product properties along with event properties

### PR DESCRIPTION
**Pull request by Innovative-GauravKochar**
This Pr is to combine the Braze custom product properties along with event properties so that Braze can map these custom product keys as well.

Jira Ticket:- https://segment.atlassian.net/jira/software/c/projects/STRATCONN/boards/371?modal=detail&selectedIssue=STRATCONN-2441&assignee=63617339fc0cc7a600b03c6b

## Testing

_Include any additional information about the testing you have completed to
ensure your changes behave as expected. For a speedy review, please check
any of the tasks you completed below during your testing._

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [x] [Segmenters] Tested in the staging environment

Screenshot 2023-04-11 at 5 00 51 PM

image

### CODE REVIEW

**Review by varadarajan-tw — COMMENTED** [review]
Good work! 🚀  Left a few suggestions.

File: packages/destination-actions/src/destinations/braze/utils.ts [hunk 1]
```diff
@@ -104,28 +104,35 @@ export function sendTrackPurchase(request: RequestClient, settings: Settings, pa
...
   const base = {
     braze_id,
     external_id,
     user_alias,
     app_id: settings.app_id,
     time: toISO8601(payload.time),
-    properties,
     _update_existing_only: payload._update_existing_only
   }
+  const products: Array<Product> = payload.properties?.products as Array<Product>
```
[the remark this answers is not in the document]
Comment (Innovative-GauravKochar) [reply packages/destination-actions/src/destinations/braze/utils.ts:117]: payload.product was actually mapped to only those properties which we defined , Any custom properties was not getting mapped .That's Why I used payload.properties.products

[the remark this answers is not in the document]
Comment (varadarajan-tw) [reply packages/destination-actions/src/destinations/braze/utils.ts:117]: oh! If thats the case, then i don't think this solution will work for all cases. What if products is not in properties? It is possible to for the customer not use the default mapping where products are not sent in properties.

File: packages/destination-actions/src/destinations/braze/utils.ts [hunk 2]
```diff
@@ -104,28 +104,34 @@ export function sendTrackPurchase(request: RequestClient, settings: Settings, pa
...
   const base = {
     braze_id,
     external_id,
     user_alias,
     app_id: settings.app_id,
     time: toISO8601(payload.time),
-    properties,
     _update_existing_only: payload._update_existing_only
   }
+  const products = payload?.products
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/braze/utils.ts:116]: products is required field right. So, is the null check needed? Can it just be payload.products?

File: packages/destination-actions/src/destinations/braze/utils.ts [hunk 3]
```diff
@@ -104,28 +104,34 @@ export function sendTrackPurchase(request: RequestClient, settings: Settings, pa
...
     ...(payload.products.length > 1 ? { headers: { 'X-Braze-Batch': 'true' } } : undefined),
     json: {
-      purchases: payload.products.map((product) => ({
-        ...base,
-        product_id: product.product_id,
-        currency: product.currency ?? 'USD',
-        price: product.price,
-        quantity: product.quantity
-      }))
+      purchases: products?.map(function (product) {
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/braze/utils.ts:122]:
```suggestion
      purchases: payload.products.map(function (product) {
```

File: packages/destination-actions/src/destinations/braze/utils.ts (hunk 2 above)
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/braze/utils.ts:116]:
```suggestion
```

**Review by varadarajan-tw — APPROVED** [review]
Thanks for doing the changes! Left couple of comments but they aren't a blocker. LGTM.

**Pull request merged**